### PR TITLE
Warn on direct state manipulation (#2272)

### LINKED ISSUE #2272: Feature request: warn when this.state is touched directly rather than with setState
I've just stumbled across some code written which touches this.state directly and then forces an update.  The danger in this pattern is that _appears_ to work, it still leaves the code vulnerable to other callers of setState doing the right thing, being queued up, and then overwriting the hand-touched state, if I'm understanding the red box on http://facebook.github.io/react/docs/component-api.html correctly.

Any chance if, in (at least) the dev version, this.state could be guarded with a JS setter that would print a warning?

### PR DESCRIPTION
**Pull request by iansu**
Adds a warning when setting state directly using this.state instead of this.setState().

Resolves #2272

### CODE REVIEW

**Comment by joh1234** [comment]
Update README
https://goo.gl/JBljdR

or 
http://bit.ly/2dK6Ojk

File: src/renderers/shared/stack/reconciler/ReactCompositeComponent.js [hunk 1]
```diff
@@ -880,6 +886,12 @@ var ReactCompositeComponent = {
         'boolean value. Make sure to return true or false.',
         this.getName() || 'ReactCompositeComponent'
       );
+
+      warning(
+        shallowEqual(this._currentState, inst.state),
+        'this.state: Setting state directly with this.state is not ' +
+        'recommended. Instead, use this.setState().'
+      );
```
Comment (iansu) [inline src/renderers/shared/stack/reconciler/ReactCompositeComponent.js]: That's fair. Adding the name of the component to the warning is relatively easy. I should have done that in the first place. Showing what was changed is definitely more difficult. If people think that's valuable then I'm willing to look into it.

**Comment by oksas** [comment]
@iansu were you still working on this?

**Comment by iansu** [comment]
@oksas I'd like to continue working on this. I have some ideas to improve the implementation, it's just unclear if those improvements will be enough to justify including this feature. Also, for some reason I didn't get a notification that @gaearon mentioned this pull request. 😞 

Possible improvements:

- Only create a copy of the component state in dev so there's no performance impact in production
- Add the name of the component to the warning message
- Show what part of the state was changed and where (I'm not exactly sure how to implement this one, so any feedback would be appreciated)

**Comment by aweary** [comment]
@iansu can you resolve the merge conflicts? This LGTM, I think we can move forward with it once its mergeable.

**Comment by iansu** [comment]
Done. Just waiting for the tests to run on CI now.

**Comment by iansu** [comment]
@aweary this is done now. Not sure if you saw my previous message because I didn't explicitly mention you.

**Comment by aweary** [comment]
@iansu thanks! Can you run the `./scripts/fiber/record-tests` script and commit the results please?

**Comment by iansu** [comment]
@aweary done! Looks like a couple of tests might be failing because of my changes. I'm not too familiar with Fiber though. Let me know if I should try and fix them.

**Comment by aweary** [comment]
@iansu sorry I'm actually not too familiar with fiber yet either, and I'm not sure if we want to implement this in fiber as well before merging. cc @gaearon for those questions

**Comment by iansu** [comment]
@aweary @gaearon sounds good. Is there any documentation on running the Fiber tests? The "How to Contribute" page just says to run `npm test`. I was a bit surprised when this first ran on CI and it ran three different sets of tests. I ended up looking at the console output there to figure out how to run the SSR tests but even that seems to have changed with the move to CircleCI. Might be a good opportunity for someone to contribute to the docs.

**Comment by bvaughn** [comment]
> Is there any documentation on running the Fiber tests?

@iansu: Sorry for chiming in late, but you can run tests against Fiber using the `REACT_DOM_JEST_USE_FIBER` env flag (eg `REACT_DOM_JEST_USE_FIBER=true jest`). Keep in mind that some failures are expected currently. The `./scripts/fiber/record-tests` will let you know if any _new_ failures were introduced because of local changes though.

That said, I think we actually want to pass on this change. Even though we warn if `this.state` is replaced, we still support it (and will likely continue to support it). I don't think it's worth the additional cost of cloning and doing a shallow comparison of `state` after each lifecycle method just to warn about something that we actually support (even though it's not recommended).

So I'm going to close this PR. Sorry for the inconvenience! And thanks a ton for the time and effort you spent on it. 🙇

**Comment by iansu** [comment]
@bvaughn Well that's disappointing but I appreciate the explanation. I'll keep looking for other opportunities to contribute. 😀

**Pull request closed without merging**